Early changes to support state recovery

j0sh opened the pull request:
Submitting now to ease review of https://github.com/livepeer/go-livepeer/pull/358, and to minimize potential merge headaches later on as the codebase continues to evolve.

Mostly consists of some helpers, the `AddReceipt` refactor, and a minor change to make `doTranscode` stand on its own, plus fixes to make tests pass.

Each commit should make sense on its own, but bundling them in the same PR for expediency.

Note that the DB changes are not in here yet, since dropping them in is fairly non-invasive at this point.

j0sh committed:
eth: Add helper to fetch latest block number.

j0sh committed:
claimmanager: Refactor AddReceipt.

Call AddReceipt once per source segment instead of incrementally
per transcoded segment, and pre-compute validity checks (eg,
profile matches) and the concatenated hashes, so we don't have
to do the computation at claim time.

This makes the claims data easier to store and recreate from DB.

j0sh committed:
test: s/common/ethcommon/g in claimmanager_test.

Improves consistency other parts of the codebase.

j0sh committed:
claimmanager: Additional tests.

ericxtang commented:
LGTM.  Like you might have mentioned in the comments - we are currently keeping the data in memory as we connect the receipts.  This is ok for the short term, but I think we'll have to evolve to either using disk or periodically claiming. Keeping data in memory can be bad for long-running streams.

ericxtang reviewed (approved):
🚢

yondonfu commented on eth/eventservices/jobservice.go:
```diff
@@ -102,8 +102,14 @@ func (s *JobService) Stop() error {
+	return DoTranscode(s.node, job)
+}
+
+func DoTranscode(node *core.LivepeerNode, job *lpTypes.Job) (bool, error) {
```
Since this is exported and callable on its own now should we kill `doTranscode` and directly invoke `DoTranscode` in the callback for `s.eventMonitor.SubscribeNewJob()` in `Start()`?

j0sh replied:
Actually, now that we're going to do all the recovery within ClaimManager, I thought it might be consistent to also handle job recovery completely within JobService (https://github.com/livepeer/go-livepeer/pull/358/commits/58ddebd62ce95b4fb3452a6188a562d97da8489a). So now we don't actually need this particular change anymore.

j0sh committed:
vendor: Update lpms to 68fe14a.

j0sh committed:
claimmanager: s/common/ethcommon/g.

Improves consistency with other parts of the codebase.

j0sh commented:
Rolling in three more commits -- a ethcommon rename in claimmanager, and vendored updates for basicnet / lpms. Note that the commit changing `doTranscode` to `DoTranscode` has been removed since it's not needed anymore.

j0sh committed:
eth: Improve logging.

claimmanager : Prefix messages with job ID where necessary.
               Additional error logging.

eventmonitor : Add event name in some error logging.

j0sh commented:
Okay, checked the effects of the potentially transaction-affecting change to `AddReceipt`, on Rinkeby. Cross checked logs, etherscan, abi-decode, etc.  Things seem OK for the most part. The first time I managed to get this working, the claims were *very* scattered, logs here: https://gist.github.com/j0sh/7d21647e047d72cb530a258486aba167 

Basically, the transcoder claimed segments 4-34, then submitted each claim individually after that up until 64, and they're all out of order. However there also seemed to be tons of network/connectivity issues during that time, many failed transactions, etc. All very strange, but it all seemed to eventually go through anyway (took a few hours). I tried again after a transcoder restart, and everything went through cleanly in one fell swoop.

Also added one more commit improving claimmanager and eventservice logging: https://github.com/livepeer/go-livepeer/pull/365/commits/3ecdf8fe16e16eb39896da528674a8246e490daa

@yondonfu @ericxtang If all this sounds good I'll go ahead and merge.

ericxtang commented:
I've seen that happen before, but never spent enough time to figure out why.  However, this logfile creates a more telling story.

One "claim" action was done here - 
https://gist.github.com/j0sh/7d21647e047d72cb530a258486aba167#file-gistfile1-txt-L1177
All of the segments were 1 apart from each other

The other "claim" was done here - 
https://gist.github.com/j0sh/7d21647e047d72cb530a258486aba167#file-gistfile1-txt-L1674
Again, a lot of the segments were 1 apart from each other

It almost feels like there were 2 ClaimManagers, each getting a part of the receipts.  It's a bad bug because it's hard to reproduce.

But in the mean time, we should consider the side effect of this problem.  Submitting many `claimWork` transactions can be really expensive for transcoders.  What do you think about setting a minimum segment # threshold, and we don't submit the claim if the count in the array is lower than the threshold?  This way - the transcoder won't claim any work in this situation, but at least it won't wasting a bunch of gas.

dob commented:
+1 for the minimum segment threshold. Bonus points if it's configurable via
a flag, but has a sensible default. It will be way too expensive to claim
individual segments. As transcoders get smarter, more configurable code,
they can decide for themselves what their policy is about when to claim, or
when to just disconnect from a job that isn't feeding them reliable
continuous segments.

On Sat, Apr 21, 2018 at 12:49 PM, Eric Tang <notifications@github.com>
wrote:

> I've seen that happen before, but never spent enough time to figure out
> why. However, this logfile creates a more telling story.
>
> One "claim" action was done here -
> https://gist.github.com/j0sh/7d21647e047d72cb530a258486aba1
> 67#file-gistfile1-txt-L1177
> All of the segments were 1 apart from each other
>
> The other "claim" was done here -
> https://gist.github.com/j0sh/7d21647e047d72cb530a258486aba1
> 67#file-gistfile1-txt-L1674
> Again, a lot of the segments were 1 apart from each other
>
> It almost feels like there were 2 ClaimManagers, each getting a part of
> the receipts. It's a bad bug because it's hard to reproduce.
>
> But in the mean time, we should consider the side effect of this problem.
> Submitting many claimWork transactions can be really expensive for
> transcoders. What do you think about setting a minimum segment # threshold,
> and we don't submit the claim if the count in the array is lower than the
> threshold? This way - the transcoder won't claim any work in this
> situation, but at least it won't wasting a bunch of gas.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/livepeer/go-livepeer/pull/365#issuecomment-383312107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAole6u_VOnmnO-YbqOasNDSAICc7CWks5tq2MbgaJpZM4TXVjN>
> .
>
